### PR TITLE
Wsteth connector

### PR DESCRIPTION
**Pull request by shriyatyagii**
Added WSTETH connector with `deposit` and `withdraw` functions.

### CODE REVIEW

File: contracts/mainnet/connectors/wsteth/main.sol [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+        uint256 getId,
+        uint256 setId
+    ) external returns (string memory _eventName, bytes memory _eventParam) {
+
+        uint256 _amt = getUint(getId, stethAmt);
+        _amt = _amt == uint(-1) ? _amt = stethContract.balanceOf(address(this)) : _amt;
+
+        approve(stethContract, address(wstethContract), _amt);
+
+        setId = wstethContract.wrap(_amt);
```
Comment (thrilok209) [inline contracts/mainnet/connectors/wsteth/main.sol]: @shriyatyagii we need to set the amount in setId and use setUnit amount.

File: contracts/mainnet/connectors/wsteth/events.sol [hunk 2]
```diff
@@ -0,0 +1,7 @@
+//SPDX-License-Identifier: MIT
+pragma solidity ^0.7.0;
+
+contract Events {
+    event LogDepositSTETH(uint256 tokenAmt, uint256 getId, uint256 setId);
```
Comment (thrilok209) [inline contracts/mainnet/connectors/wsteth/events.sol]: @shriyatyagii we can just keep LogDeposit and LogWithdraw

File: contracts/mainnet/connectors/wsteth/main.sol (hunk 1 above)
Comment (shriyatyagii) [reply contracts/mainnet/connectors/wsteth/main.sol:31]: Done

File: contracts/mainnet/connectors/wsteth/events.sol (hunk 2 above)
Comment (shriyatyagii) [reply contracts/mainnet/connectors/wsteth/events.sol]: Done

File: contracts/mainnet/connectors/wsteth/main.sol (hunk 1 above)
Comment (shriyatyagii) [reply contracts/mainnet/connectors/wsteth/main.sol]: Done

File: contracts/mainnet/connectors/wsteth/events.sol [hunk 3]
```diff
@@ -0,0 +1,7 @@
+//SPDX-License-Identifier: MIT
+pragma solidity ^0.7.0;
+
+contract Events {
+    event LogDeposit(uint256 tokenAmt, uint256 getId, uint256 setId);
+    event LogWithdraw(uint256 tokenAmt, uint256 getId, uint256 setId);
```
Comment (thrilok209) [inline contracts/mainnet/connectors/wsteth/events.sol]: @shriyatyagii I think we can emit both wsteth and steth amount?

File: contracts/mainnet/connectors/wsteth/main.sol [hunk 4]
```diff
@@ -0,0 +1,64 @@
...
+        uint256 wstethAmt,
+        uint256 getId,
+        uint256 setId
+    ) external returns (string memory _eventName, bytes memory _eventParam) {
+
+        uint256 _amt = getUint(getId, wstethAmt);
+        _amt = _amt == uint(-1) ? wstethContract.balanceOf(address(this)) : _amt;
+
+        wstethContract.unwrap(_amt);
+        setUint(setId, _amt);
```
Comment (thrilok209) [inline contracts/mainnet/connectors/wsteth/main.sol]:
```suggestion
        uint256 stethAmt_ = wstethContract.unwrap(_amt);
        setUint(setId, stethAmt_);
```

File: contracts/mainnet/connectors/wsteth/main.sol [hunk 5]
```diff
@@ -0,0 +1,64 @@
...
+        uint256 setId
+    ) external returns (string memory _eventName, bytes memory _eventParam) {
+
+        uint256 _amt = getUint(getId, stethAmt);
+        _amt = _amt == uint(-1) ? _amt = stethContract.balanceOf(address(this)) : _amt;
+
+        approve(stethContract, address(wstethContract), _amt);
+
+        wstethContract.wrap(_amt);
+        setUint(setId, _amt);
```
Comment (thrilok209) [inline contracts/mainnet/connectors/wsteth/main.sol]:
```suggestion
        uint256 wstethAmt_ = wstethContract.wrap(_amt);
        setUint(setId, wstethAmt_);
```

File: contracts/mainnet/connectors/wsteth/main.sol [hunk 6]
```diff
@@ -0,0 +1,64 @@
...
+
+        uint256 _amt = getUint(getId, stethAmt);
+        _amt = _amt == uint(-1) ? _amt = stethContract.balanceOf(address(this)) : _amt;
+
+        approve(stethContract, address(wstethContract), _amt);
+
+        uint256 wstethAmt_ = wstethContract.wrap(_amt);
+        setUint(setId, wstethAmt_);
+
+        _eventName = "LogDeposit(uint256,uint256,uint256)";
```
Comment (thrilok209) [inline contracts/mainnet/connectors/wsteth/main.sol]:
```suggestion
        _eventName = "LogDeposit(uint256,uint256,uint256,uint256)";
```

File: contracts/mainnet/connectors/wsteth/main.sol [hunk 7]
```diff
@@ -0,0 +1,64 @@
...
+        uint256 setId
+    ) external returns (string memory _eventName, bytes memory _eventParam) {
+
+        uint256 _amt = getUint(getId, wstethAmt);
+        _amt = _amt == uint(-1) ? wstethContract.balanceOf(address(this)) : _amt;
+
+        uint256 stethAmt_ = wstethContract.unwrap(_amt);
+        setUint(setId, stethAmt_);
+
+        _eventName = "LogWithdraw(uint256,uint256,uint256)";
```
Comment (thrilok209) [inline contracts/mainnet/connectors/wsteth/main.sol]:
```suggestion
        _eventName = "LogWithdraw(uint256,uint256,uint256,uint256)";
```

File: contracts/mainnet/connectors/wsteth/main.sol [hunk 8]
```diff
@@ -0,0 +1,64 @@
...
+        uint256 setId
+    ) external returns (string memory _eventName, bytes memory _eventParam) {
+
+        uint256 _amt = getUint(getId, stethAmt);
+        _amt = _amt == uint(-1) ? _amt = stethContract.balanceOf(address(this)) : _amt;
+
+        approve(stethContract, address(wstethContract), _amt);
+
+        uint256 wstethAmt_ = wstethContract.wrap(_amt);
+        setUint(setId, wstethAmt_);
```
Comment (thrilok209) [inline contracts/mainnet/connectors/wsteth/main.sol]:
```suggestion
        uint256 _wstethAmt = wstethContract.wrap(_amt);
        setUint(setId, _wstethAmt);
```

File: contracts/mainnet/connectors/wsteth/main.sol [hunk 9]
```diff
@@ -0,0 +1,64 @@
...
+        uint256 wstethAmt,
+        uint256 getId,
+        uint256 setId
+    ) external returns (string memory _eventName, bytes memory _eventParam) {
+
+        uint256 _amt = getUint(getId, wstethAmt);
+        _amt = _amt == uint(-1) ? wstethContract.balanceOf(address(this)) : _amt;
+
+        uint256 stethAmt_ = wstethContract.unwrap(_amt);
+        setUint(setId, stethAmt_);
```
Comment (thrilok209) [inline contracts/mainnet/connectors/wsteth/main.sol]:
```suggestion
        uint256 _stethAmt = wstethContract.unwrap(_amt);
        setUint(setId, _stethAmt);
```

File: contracts/mainnet/connectors/wsteth/main.sol [hunk 10]
```diff
@@ -0,0 +1,64 @@
...
+*/
+
+abstract contract WSTETHContract is Helpers, Basic {
+
+    /**
+     * @dev Deposit STETH into WSTETH.
+     * @notice Wrap STETH into WSTETH
+     * @param stethAmt The amount of STETH to deposit. (For max: `uint256(-1)`)
+     * @param getId ID to retrieve amt.
+     * @param setId ID stores the amount of STETH deposited.
```
Comment (thrilok209) [inline contracts/mainnet/connectors/wsteth/main.sol]:
```suggestion
     * @param getId ID to retrieve STETH amount.
     * @param setId ID stores the amount of WSTETH deposited.
```

File: contracts/mainnet/connectors/wsteth/main.sol [hunk 11]
```diff
@@ -0,0 +1,64 @@
...
+        _eventName = "LogDeposit(uint256,uint256,uint256,uint256)";
+        _eventParam = abi.encode(_amt, wstethAmt_, getId, setId);
+    }
+
+    /**
+     * @dev Withdraw STETH from WSTETH from Smart  Account
+     * @notice Unwrap STETH from WSTETH
+     * @param wstethAmt The amount of WSTETH to withdraw. (For max: `uint256(-1)`)
+     * @param getId ID to retrieve amt.
+     * @param setId ID stores the amount of ETH withdrawn.
```
Comment (thrilok209) [inline contracts/mainnet/connectors/wsteth/main.sol]:
```suggestion
     * @param getId ID to retrieve WSTETH amount.
     * @param setId ID stores the amount of STETH.
```

File: contracts/mainnet/connectors/wsteth/main.sol [hunk 12]
```diff
@@ -0,0 +1,64 @@
...
+    ) external returns (string memory _eventName, bytes memory _eventParam) {
+
+        uint256 _amt = getUint(getId, wstethAmt);
+        _amt = _amt == uint(-1) ? wstethContract.balanceOf(address(this)) : _amt;
+
+        uint256 _stethAmt = wstethContract.unwrap(_amt);
+        setUint(setId, _stethAmt);
+
+        _eventName = "LogWithdraw(uint256,uint256,uint256,uint256)";
+        _eventParam = abi.encode(_amt, stethAmt_, getId, setId);
```
Comment (thrilok209) [inline contracts/mainnet/connectors/wsteth/main.sol]:
```suggestion
        _eventParam = abi.encode(_amt, _stethAmt, getId, setId);
```

File: contracts/mainnet/connectors/wsteth/main.sol [hunk 13]
```diff
@@ -0,0 +1,64 @@
...
+        uint256 _amt = getUint(getId, stethAmt);
+        _amt = _amt == uint(-1) ? _amt = stethContract.balanceOf(address(this)) : _amt;
+
+        approve(stethContract, address(wstethContract), _amt);
+
+        uint256 _wstethAmt = wstethContract.wrap(_amt);
+        setUint(setId, _wstethAmt);
+
+        _eventName = "LogDeposit(uint256,uint256,uint256,uint256)";
+        _eventParam = abi.encode(_amt, wstethAmt_, getId, setId);
```
Comment (thrilok209) [inline contracts/mainnet/connectors/wsteth/main.sol]:
```suggestion
        _eventParam = abi.encode(_amt, _wstethAmt, getId, setId);
```

**Comment by shriyatyagii** [comment]
Connector details -
Name - `WSTETH-A`
Address - [0xa1040eaaB1D5ba731aAD6D9b59C70E1E1a5BEd57](https://etherscan.io/address/0xa1040eaaB1D5ba731aAD6D9b59C70E1E1a5BEd57#code)

Transaction - [Tx](https://etherscan.io/tx/0x0f1e1b46bf7980170b79d21f10dce2ea347d6fbc6a4765f70800af4bf829e6ad)

**Review by thrilok209 — APPROVED** [review]

**Pull request merged**